packages/react-ui-components/src/components/SchemaForm/SchemaForm.tsx

```diff
@@ -0,0 +1,111 @@
+import { EActionButtonType, EComponentSize } from '@kelvininc/ui-components';
+import Form, { FormProps, IChangeEvent, utils, withTheme } from '@rjsf/core';
+import classNames from 'classnames';
+import { cloneDeep, isEmpty, isEqualWith } from 'lodash-es';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import { useScroll } from '../../utils';
+import { KvActionButtonText } from '../stencil-generated';
+import styles from './SchemaForm.module.scss';
+import Theme from './Theme';
+
+const ThemedForm = withTheme(Theme);
+const { getSubmitButtonOptions } = utils;
+const SCROLL_OFFSET = 10;
+
+interface SchemaForm<T> extends Form<T> {
+	formElement: HTMLDivElement;
+}
+
+export interface SchemaFormProps<T> extends FormProps<T> {
+	submittedData?: T;
+	allowDiscardChanges?: boolean;
+}
+
+export function KvSchemaForm<T>(props: SchemaFormProps<T>) {
+	const [isValid, setValid] = useState(!props.liveValidate);
+	const [hasChanges, setHasChanges] = useState(!isEqualWith(props.formData, props.submittedData || {}));
+	const [formData, setFormData] = useState(props.formData || ({} as T));
+	const [templateElem, setTemplateElem] = useState<HTMLDivElement | undefined>(undefined);
+
+	const formRef = useRef<SchemaForm<T>>(null);
+	const { scrollTop } = useScroll(templateElem);
+	const isScrolling = useMemo(() => scrollTop - SCROLL_OFFSET > 0, [scrollTop]);
+	const { submitText, norender, props: submitButtonProps } = getSubmitButtonOptions(props.uiSchema || {});
+	const hasFooter = props.allowDiscardChanges || !norender;
+
+	const themedProps = {
+		...props,
+		ref: formRef,
+		onChange: (data: IChangeEvent<T>) => {
+			const hasNewChanges = !isEqualWith(data.formData || {}, props.submittedData || {});
```

Comment:
Can you use data destructuring with default value of `{}`?

Reply:
Please don't forget this 😄

Reply:
yes you can have the `submitedData` equals to `undefined` at the first instant and `formData` equals to `{}` and in that case don't exists changes yet.